merkle/compact/tree.go

```diff
@@ -94,35 +92,39 @@ func NewTreeWithState(hasher hashers.LogHasher, size int64, getNodesFn GetNodesF
 	for i, id := range ids {
 		r.nodes[id.Level] = hashes[i]
 	}
-	r.recalculateRoot(func(NodeID, []byte) {})
 
-	if !bytes.Equal(r.root, expectedRoot) {
-		glog.Warningf("Corrupt state, expected root %s, got %s", hex.EncodeToString(expectedRoot[:]), hex.EncodeToString(r.root[:]))
-		return nil, fmt.Errorf("root hash mismatch: got %v, expected %v", r.root, expectedRoot)
+	// TODO(pavelkalinnikov): This check should be done externally.
+	root, err := r.CurrentRoot()
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(root, expectedRoot) {
+		glog.Warningf("Corrupt state, expected root %s, got %s", hex.EncodeToString(expectedRoot[:]), hex.EncodeToString(root))
+		return nil, fmt.Errorf("root hash mismatch: got %v, expected %v", root, expectedRoot)
 	}
-	glog.V(1).Infof("Resuming at size %d, with root: %s", r.size, base64.StdEncoding.EncodeToString(r.root[:]))
+
+	glog.V(1).Infof("Resuming at size %d, with root: %s", r.size, base64.StdEncoding.EncodeToString(root))
 	return &r, nil
 }
 
 // NewTree creates a new compact Tree with size zero.
 func NewTree(hasher hashers.LogHasher) *Tree {
 	return &Tree{
 		hasher: hasher,
-		root:   hasher.EmptyRoot(),
 		nodes:  make([][]byte, 0),
 		size:   0,
 	}
 }
 
 // CurrentRoot returns the current root hash.
-func (t *Tree) CurrentRoot() []byte {
-	return t.root
+func (t *Tree) CurrentRoot() ([]byte, error) {
+	return t.calculateRoot(func(NodeID, []byte) {})
```

Comment:
This makes this function more expensive, are there any existing callers who assume that it isn't?

Reply:
I didn't find any dramatic ones. Usage patterns that I found:

1. Call `CurrentRoot` after every `AddLeafHash`. Only in tests.
2. Call it after a series of `AddLeafHash`, like in sequencer or tree cache.
3. Call it once, e.g. during serialization/deserialization in the ingestor.

We can ignore case 1 because it's tests, and case 3 is not affected. In case 2 there will be one extra `calculateRoot` call in the whole sequence (so N+1 calls instead of N, for N leaves added). But, see the "bigger picture" comment below, in a follow-up PR N+1 will be be decreased to 1 (`calculateRoot` will be called "lazily" after all inserts are completed).